### ISSUE TITLE
One last VT color issue in 1.2 - transparent foreground (default background while reversed)

### ISSUE THREAD

**Issue by PhMajerus** [microsoft/terminal#7014]
The new Preview 1.2.2022.0 seems to fix all the VT colors issues I observed before, except one last case:

When the background color is the default (49), it is correctly kept transparent, but if the foreground color is transparent, which is possible when the background is 49 and then the colors are reversed using 7, it is rendered as black instead of transparent.

![image](https://user-images.githubusercontent.com/25664275/88215128-7c6f9880-cc5b-11ea-98bc-68d86772af70.png)
echo -e "\e[31m\u2580\e[7m\u2580\e[m"
echo -e "\e[31m\u2590\e[7m\u25A0\e[27m\u258C\e[m"

In both cases, there should be no black square, the acrylic transparency should show through instead.

**Comment by DHowett** [comment]
Oh yeah, we need to do an inverse alpha mask with the shaped text. It's expensive and something we chose to not to.

**Comment by PhMajerus** [comment]
Could we get that behavior through an optional flag ?
I understand your performance goal, but for some ANSI-art effects, the only solution is to use an existing character reversed, such as for the 3-cells square above. Would be nice to be able to turn it on for specific profiles.

**Comment by DHowett** [comment]
Meh, I'd be inclined to implement it and not make it configurable -- it's more expensive on the engineering side than it is on the runtime side, I expect. :smile:

**Comment by PhMajerus** [comment]
@DHowett 👍 Great, can't wait to see it implemented then 😁

Especially since you can use it only when you have that specific attributes combination, the current code can be used for faster rendering in every other case, and that combination would definitely look more correct taking advantage of the acrylic effect.

**Comment by DHowett** [comment]
Notes:

We can't use a geometry mask based on the glyph run, because we want to fill _outside_ of the geometry and not touch _inside_ the geometry[1].

We may need to use an alpha bitmap (1bpp black and white, converted to an alpha mask) like we did in [Islandwood's CGImage](https://github.com/microsoft/WinObjC/blob/ea3f7983803fa02309f974ff878b6c9ecc72c7c4/Frameworks/CoreGraphics/CGImage.mm#L805-L864) and push a layer whose opacity bitmap is this guy.

[1]: Nevermind, I just saw [ID2D1Geometry::CombineWithGeometry](https://docs.microsoft.com/en-us/windows/win32/api/d2d1/nf-d2d1-id2d1geometry-combinewithgeometry(id2d1geometry_d2d1_combine_mode_constd2d1_matrix_3x2_f__id2d1simplifiedgeometrysink)?redirectedfrom=MSDN); perhaps we can just create a Rectangle geometry, combine EXCLUDE the glyph geometries, and fill that...

**Comment by thanhph111** [comment]
I understand there're more priority to get done but I just want to add another case what need this feature more frequently.
As you can see in the image below, I use [posh-git](https://github.com/dahlbyk/posh-git) to draw prompt with segment forward symbol expected to be colored like background. However, as `useAcrylic` is set, the symbol is revealed with default black.
![image](https://user-images.githubusercontent.com/60842560/88313264-7a7e0600-cd3d-11ea-8b0f-e53ae53cb6bc.png)
Hope to see this one added in the future.

**Comment by DHowett** [comment]
@thanhph111 

Unless you're using reverse video and your powerline prompt is actually using `49` as the background (not `40`), "...LAPTOP" is supposed to be white on black.

Can you confirm whether your prompt is officially using `7;49;32` (blue on default, reverse) or `42;30` (black on blue)?

**Comment by thanhph111** [comment]
@DHowett 
I actually get the color from the console and use the `Write-Prompt` (code below) to draw it out. So I think issue is from the terminal. 
Maybe I miss something, please correct me.
![image](https://user-images.githubusercontent.com/60842560/88321533-88398880-cd49-11ea-8f48-6c65683f4d1a.png)
`Write-Prompt`:
``` shell
function Write-Prompt($Object, $ForegroundColor = $null, $BackgroundColor = $null) {
    $s = $global:GitPromptSettings

    if ($BackgroundColor -is [string]) {
        $BackgroundColor = [ConsoleColor]$BackgroundColor
    }
    if ($ForegroundColor -is [string]) {
        $ForegroundColor = [ConsoleColor]$ForegroundColor
    }

    $writeHostParams = @{
        Object = $Object;
        NoNewLine = $true;
    }
    if (($BackgroundColor -ge 0) -and ($BackgroundColor -le 15)) {
        $writeHostParams.BackgroundColor = $BackgroundColor
    }
    if (($ForegroundColor -ge 0) -and ($ForegroundColor -le 15)) {
        $writeHostParams.ForegroundColor = $ForegroundColor
    }
    Write-Host @writeHostParams
}
```

**Comment by DHowett** [comment]
In that case, you're seeing the results of the color compatibility work done in #6810 and your issue is not that the arrow is black but that the text to the left of it is _not_ white-on-black.

**Comment by DHowett** [comment]
Hey @PhMajerus,

![image](https://user-images.githubusercontent.com/189190/88325582-1d9d3400-ccda-11ea-9951-7bbeedad11b6.png)

![image](https://user-images.githubusercontent.com/189190/88325933-61903900-ccda-11ea-9991-88f2ae5533f5.png)

**Comment by PhMajerus** [comment]
That was fast, thank you @DHowett  \(^_^)/

**Comment by DHowett** [comment]
So, it's still got some issues. I want to make sure it works on non-native-scale displays, and follows the antialiasing mode set for the font, and uh it completely breaks hinting (because we're just getting the text geometry and not letting the rasterizer at it)

**Comment by thanhph111** [comment]
@DHowett Sorry to bother you again and mess up this issue.
> ![image](https://user-images.githubusercontent.com/189190/88325582-1d9d3400-ccda-11ea-9951-7bbeedad11b6.png)
> ![image](https://user-images.githubusercontent.com/189190/88325933-61903900-ccda-11ea-9991-88f2ae5533f5.png)

How can you do that? I try to use `Write-Host` to write the same but it didn't work. I spent all night to read #6810 as you mention but also have no clue. I did:
- [x] Run on Windows Terminal 1.2.2022.0
- [x] Run with PSReadline 2.0.3
- [x] Run on PowerShell Core 7.0.3

The test below I use black as the same as the background color (#000000) and the white color as the same as the foreground color (#FFFFFF) but it seems that `abc` is not transparent at all. I hope if text works, [glyph](https://github.com/microsoft/terminal/issues/7014#issuecomment-663117947) should work also.
![image](https://user-images.githubusercontent.com/60842560/88378704-b8297000-cdcb-11ea-9cbc-e5406324b363.png)

**Comment by PhMajerus** [comment]
@thanhph111 Write-Host arguments were designed for the Console API, they do not translate to the VT sequences we're using here.
You need to generate your VT control sequences manually, for example :
PowerShell 7 and later:
``Write-Output "`e[97;7m Transparent on white `e[m"``
Windows PowerShell:
``Write-Output "$([char]27)[97;7m Transparent on white $([char]27)[m"``
Or you can use some other cmdlet or component that can generate strings with proper escape sequences.

Of course these will show black on white in the current 1.2 preview build of Windows Terminal, which is the topic of this issue thread.

![image](https://user-images.githubusercontent.com/25664275/88383029-fb72e700-cda9-11ea-9378-8c1731171055.png)
VT and ANSI-art in Windows PowerShell 5.1.19041.1 within Windows Terminal 1.2.2022.0.

**Comment by PhMajerus** [comment]
> So, it's still got some issues. I want to make sure it works on non-native-scale displays, and follows the antialiasing mode set for the font, and uh it completely breaks hinting (because we're just getting the text geometry and not letting the rasterizer at it)

I didn't go see your code, but since there are only two colors, and in this scenario one of them is transparent, wouldn't bit-fiddling from the positive rendering be enough (and fast if you can run that in a GPU shader) ?

It seems like your final pixels can be taken from the positive rendered cell: the positive rendered alpha reversed (A2 = 255-A1) and RGB fixed to the non transparent color for the whole cell. The alpha should be all that is needed to recover the exact opposite of the shape and RGB values of transparent pixels colors won't matter, so fixing the color for all pixels except for their alpha values should be faster and more reliable than trying to reverse the RGB values.

**Comment by DHowett** [comment]
So for efficiency's sake, I'm trying to avoid capturing the positive render as an image and then inverting it before it goes out. What we _really_ have is a set of possibly-GPU-managed buffers that we can't get access to, and the ability to push geometries to them.

There's a world where we could easily render the text with grayscale AA on white to a temporary bitmap target then draw through that as an alpha mask (ala CGCreateMaskImage from WinObjC above^), but going there would slay performance and introduce a lot of buffer/target churn.

We're also not set up to apply pixel shaders generically, and for doing it on the GPU I'm not sure we can easily grab the backing buffer and alpha-invert just the region we care about without jumping through rather a couple hoops, either.

For the best bang-for-your-buck tradeoff I'm most comfortable with the negative geometry solution... even if it does break hinting.

**Comment by jales** [comment]
@DHowett Are there any plans to merge your changes in a future version?

**Comment by DHowett** [comment]
@jales not as yet; they are insufficient right now

**Comment by Zx-EvM** [comment]
The issue is still reproduced on the latest preview version [1.15.18620](https://github.com/microsoft/terminal/releases/tag/v1.15.1862.0)

Related discussion: https://github.com/JanDeDobbeleer/oh-my-posh/discussions/2555

How to fix the issue as it done in this comment?: https://github.com/microsoft/terminal/issues/7014#issuecomment-663464319

**Comment by zeratax** [comment]
I can still reproduce this as well on 1.17.1023

**Comment by DHowett** [comment]
Thanks for the report! That would be because it has not yet been fixed.

**Comment by FrostKiwi** [comment]
Want to confirm it being broken on my machine as well with `ver 1.17.11461.0`, as reported in https://github.com/microsoft/terminal/issues/15491 and https://github.com/microsoft/terminal/issues/15491
![image](https://github.com/microsoft/terminal/assets/60887273/f9a474a7-4226-4205-8d5d-741adc0cd798)

**Comment by DHowett** [comment]
Thanks for the report! Yes, this is not machine-specific and requires explicit work in the renderer. We recently put a lot of effort into the new rendering engine, and we _think_ that it should be possible to do this properly.

**Comment by silkfire** [comment]
What's the status on this? Still present in 1.18.3181.0

![image](https://github.com/microsoft/terminal/assets/1554615/56e4dcf4-08aa-4088-8646-bac5ec8e5adc)

**Comment by zadjii-msft** [comment]
Yep. We'll make sure to update this thread when there are updates. If someone's interested in trying to fix this themselves, we'd be more than happy to point in the right direction!

**Comment by glitchplaysgames714** [comment]
is this still being worked on?? having this issue aswell

**Comment by j4james** [comment]
FYI, there is an [inverse variant](https://github.com/ryanoasis/nerd-fonts/pull/1490) of the arrow head glyphs that was added to nerd fonts a couple of years ago, which provides a way to work around this transparency problem. For more info, see https://github.com/microsoft/terminal/issues/18266#issuecomment-2718722011.

**Comment by pfmoore** [comment]
> which provides a way to work around this transparency problem

It doesn't work in all situations, and even when it does, it requires extra complication in your setup to use the additional characters. So getting this issue fixed would still be worthwhile.

**Comment by glitchplaysgames714** [comment]
> > which provides a way to work around this transparency problem
> 
> It doesn't work in all situations, and even when it does, it requires extra complication in your setup to use the additional characters. So getting this issue fixed would still be worthwhile.

yea it doesnt work most of the time. it works fine in powershell but vscode doesnt like it and makes a gap

<img width="1810" height="260" alt="Image" src="https://github.com/user-attachments/assets/d6558fec-be30-4546-9ec8-810a8c705369" />

<img width="249" height="37" alt="Image" src="https://github.com/user-attachments/assets/1f6d7134-d3fe-4ab4-abb6-1a3cbf962d3a" />

**Comment by pfmoore** [comment]
Does VS Code have a different rendering engine? It might be necessary to raise a separate issue with them - I honestly don't know 🙁 It would be nice if the problem could be solved in a single place, but I don't know if that's the reality here.

**Comment by lhecker** [comment]
Indeed, VS Code has its own terminal implementation called `xterm.js`.
